File: data-generators/headers/converter.cpp
#include "converter.h"

std::vector<long long> factorial;

// Ordered Statistics Tree
struct OrderedSet {
  std::vector<int> BIT;
  int sz;

  OrderedSet(int n) {
    sz = 1;
    while (sz < n) {
      sz <<= 1;
    }
    BIT.resize(sz+1, 0);
  }

  void insert(int x) {
    while (x <= sz) {
      BIT[x] += 1;
      x += (x&-x);
    }
  }

  void erase(int x) {
    while (x <= sz) {
      BIT[x] -= 1;
      x += (x&-x);
    }
  }

  int order_of_key(int x) {
    int order = 0;
    while (x > 0) {
      order += BIT[x];
      x -= (x&-x);
    }
    return order-1;
  }

  int find_by_order(int k) {
    int order = 0;
    int result = 0;
    for (int i = sz; i > 0; i >>= 1) {
      if (order + BIT[i+result] <= k) {
        order += BIT[i+result];
        result += i;
      }
    }

    return result+1;
  }
};

std::vector<int> get_permutation_from_index(long long k, int n) {
  std::vector<int> p(n);
  OrderedSet ordered_set(n);
  for (int i = 0; i < n; i++) ordered_set.insert(i+1);

  if (factorial.size() < n) {
    factorial.resize(n);
    factorial[0] = 1;
    for (int i = 1; i < n; i++) factorial[i] = i*factorial[i-1];
  }

  for (int i = 0; i < n; i++) {
    int lo = 0, hi = n-i-1, ans = 0;
    while (lo <= hi) {
      int mid = hi - (hi-lo)/2;
      if (mid*factorial[n-i-1] <= k) {
        ans = mid;
        lo = mid+1;
      }
      else hi = mid-1;
    }

    p[i] = ordered_set.find_by_order(ans);
    k -= ans*factorial[n-i-1];
    ordered_set.erase(p[i]);
  }
  return p;
}

long long get_index_of_permutation(std::vector<int> &p) {
  int n = p.size();
  OrderedSet ordered_set(n);
  for (int i = 0; i < n; i++) ordered_set.insert(i+1);

  if (factorial.size() < n) {
    factorial.resize(n);
    factorial[0] = 1;
    for (int i = 1; i < n; i++) factorial[i] = i*factorial[i-1];
  }

  long long index = 0;
  for (int i = 0; i < n; i++) {
    index += ordered_set.order_of_key(p[i])*factorial[n-i-1];
    ordered_set.erase(p[i]);
  }
  return index;
}

File: supervised-learning/converter.py
import numpy as np

class OrderedSet():
  def __init__(self, n):
    self.size = 1
    while self.size < n: 
      self.size <<= 1

    self.BIT = [0]*(self.size+1)

  def insert(self, x):
    while x <= self.size:
      self.BIT[x] += 1
      x += (x&-x)
  
  def erase(self, x):
    while x <= self.size:
      self.BIT[x] -= 1
      x += (x&-x)
    
  def order_of_key(self, x):
    result = 0
    while x > 0:
      result += self.BIT[x]
      x -= (x&-x)
    
    return result-1
  
  def find_by_order(self, k):
    x = self.size
    total = 0
    result = 0

    while x > 0:
      if total + self.BIT[result+x] <= k:
        total += self.BIT[result+x]
        result += x
      x >>= 1
    
    return result+1

def get_permutation_from_index(k, n):
  """
  Returns the size-n permuation of a given index

  Parameters:
    k (int): The index of the permutation (0 <= k < n!)
    n (int): The size of the permutation
  
  Returns:
    numpy.ndarray[int]: The resulting permutation
  """
  p = [0]*n
  ordered_set = OrderedSet(n)
  for i in range(n):
    ordered_set.insert(i+1)

  for i in range(n):
    lo = 0
    hi = n-i-1
    ans = 0

    while lo <= hi:
      mid = hi - (hi-lo)//2
      if mid*np.math.factorial(n-i-1) <= k:
        ans = mid
        lo = mid+1
      else:
        hi = mid-1

    p[i] = ordered_set.find_by_order(ans)
    k -= ans*np.math.factorial(n-i-1)
    ordered_set.erase(p[i])

  p = np.array(p, dtype=np.int32)
  return p
  
def get_index_of_permutation(state):
  """
  Returns the index of a permutation

  Parameters:
    state (numpy.ndarray[int]): The permutation
  
  Returns:
    int: The index of the permutaton
  """
  n = len(state)
  ordered_set = OrderedSet(n)
  for i in range(n):
    ordered_set.insert(i+1)

  index = 0
  for i in range(n):
    index += ordered_set.order_of_key(state[i])*np.math.factorial(n-i-1)
    ordered_set.erase(state[i])
  return index